presto-spi/src/main/java/io/prestosql/spi/block/Int96ArrayBlockEncoding.java

```diff
@@ -38,11 +40,25 @@ public void writeBlock(BlockEncodingSerde blockEncodingSerde, SliceOutput sliceO
 
         encodeNullsAsBits(sliceOutput, block);
 
-        for (int position = 0; position < positionCount; position++) {
-            if (!block.isNull(position)) {
-                sliceOutput.writeLong(block.getLong(position, 0));
-                sliceOutput.writeInt(block.getInt(position, 8));
+        if (!block.mayHaveNull()) {
+            sliceOutput.writeBytes(getHighSlice(block));
+            sliceOutput.writeBytes(getLowSlice(block));
+        }
+        else {
+            long[] high = new long[positionCount];
+            int[] low = new int[positionCount];
+            int nonNullPositionCount = 0;
+            for (int i = 0; i < positionCount; i++) {
+                high[nonNullPositionCount] = block.getLong(i, 0);
+                low[nonNullPositionCount] = block.getInt(i, 8);
+                if (!block.isNull(i)) {
```

Comment:
Do I remember correctly that this check before previous two lines performs worse?

Reply:
Mostly yes. It is a bit faster for datasets with a lot of nulls (>81% to be precise), but other than that it is much slower